lib/runtime-c-api/src/metering.rs

```diff
@@ -0,0 +1,111 @@
+
+/// Creates a new Module with gas limit from the given wasm bytes.
+///
+/// Returns `wasmer_result_t::WASMER_OK` upon success.
+///
+/// Returns `wasmer_result_t::WASMER_ERROR` upon failure. Use `wasmer_last_error_length`
+/// and `wasmer_last_error_message` to get an error message.
+#[allow(clippy::cast_ptr_alignment)]
+#[cfg(feature = "metering")]
+#[no_mangle]
+pub unsafe extern "C" fn wasmer_compile_with_limit(
+    module: *mut *mut wasmer_module_t,
+    wasm_bytes: *mut u8,
+    wasm_bytes_len: u32,
+    gas_limit: u64,
+) -> wasmer_result_t {
+    let bytes: &[u8] = slice::from_raw_parts_mut(wasm_bytes, wasm_bytes_len as usize);
+    let result = wasmer_runtime_core::compile_with(bytes, &get_metered_compiler(gas_limit));
+    let new_module = match result {
+        Ok(instance) => instance,
+        Err(error) => {
+            update_last_error(error);
+            return wasmer_result_t::WASMER_ERROR;
+        }
+    };
+    *module = Box::into_raw(Box::new(new_module)) as *mut wasmer_module_t;
+    wasmer_result_t::WASMER_OK
+}
+
+#[cfg(feature = "metering")]
+fn get_metered_compiler(limit: u64) -> impl Compiler {
+    use wasmer_middleware_common::metering;
+    use wasmer_runtime_core::codegen::{MiddlewareChain, StreamingCompiler};
+    use wasmer_singlepass_backend::ModuleCodeGenerator as SinglePassMCG;
+    let c: StreamingCompiler<SinglePassMCG, _, _, _, _> = StreamingCompiler::new(move || {
+        let mut chain = MiddlewareChain::new();
+        chain.push(metering::Metering::new(limit));
+        chain
+    });
+    c
+}
+
+// returns gas used
+#[allow(clippy::cast_ptr_alignment)]
+#[no_mangle]
+#[cfg(feature = "metering")]
+pub unsafe extern "C" fn wasmer_instance_get_points_used(instance: *mut wasmer_instance_t) -> u64 {
+    use wasmer_middleware_common::metering;
+    let instance = &*(instance as *const wasmer_runtime::Instance);
+    let points = metering::get_points_used(instance);
+    points
+}
+
+// sets gas used
+#[allow(clippy::cast_ptr_alignment)]
+#[no_mangle]
+#[cfg(feature = "metering")]
+pub unsafe extern "C" fn wasmer_instance_set_points_used(
+    instance: *mut wasmer_instance_t,
+    new_gas: u64,
+) {
+    use wasmer_middleware_common::metering;
+    let instance = &mut *(instance as *mut wasmer_runtime::Instance);
+    metering::set_points_used(instance, new_gas)
+}
+
+/*** placeholder implementation if metering feature off ***/
+
+// Without metering, wasmer_compile_with_limit is a copy of wasmer_compile
+#[cfg(not(feature = "metering"))]
+#[no_mangle]
+pub unsafe extern "C" fn wasmer_compile_with_limit(
+    module: *mut *mut wasmer_module_t,
+    wasm_bytes: *mut u8,
+    wasm_bytes_len: u32,
+    _: u64,
+) -> wasmer_result_t {
+    let bytes: &[u8] = slice::from_raw_parts_mut(wasm_bytes, wasm_bytes_len as usize);
+    // TODO: this implicitly uses default_compiler() is that proper? maybe we override default_compiler
+    let result = wasmer_runtime::compile(bytes);
```

Comment:
You should coerce to compile with the singlepass backend here for the sake of consistent behaviors.

Reply:
Fair enough.
Or should we force the default compiler to single-pass when metering is on, with modified feature flags? Then we can just use default_compiler() consistently everywhere, which probably makes the cleanest code.

When cranelift supports metering, then it swaps out to that as default.

This approach would mean that builds with metering feature will be slower than the other ones, so it should remain a default off feature until cranelift support. But, it would allow me to complete the go bindings and integrations already.

Given the runtime performance of the wasm is not a major issue in most expected smart contracts until we do signature verification or heavy hashing in wasm, singlepass should work great for v1 of the cosmos-sdk integration. And if cranelift+metering work is on your roadmap for the next, say, 6 months, this should be plenty fine for the downstream updates (which would be seemless).

Reply:
We always use default compiler now when metering is enabled.
This is hardcoded as llvm now, but could be changed.